sql/hive-thriftserver/v2.3.5/src/main/scala/org/apache/spark/sql/hive/thriftserver/ThriftserverShimUtils.scala

```diff
@@ -56,6 +56,12 @@ private[thriftserver] object ThriftserverShimUtils {
 
   private[thriftserver] def toJavaSQLType(s: String): Int = Type.getType(s).toJavaSQLType
 
+  private[thriftserver] def supportedType(): Seq[Type] = {
+    Array(Type.NULL_TYPE, Type.BOOLEAN_TYPE, Type.TINYINT_TYPE, Type.SMALLINT_TYPE, Type.INT_TYPE,
+      Type.BIGINT_TYPE, Type.FLOAT_TYPE, Type.DOUBLE_TYPE, Type.STRING_TYPE, Type.DATE_TYPE,
+      Type.TIMESTAMP_TYPE, Type.DECIMAL_TYPE, Type.BINARY_TYPE)
+  }
```

Comment:
Why do we skip `ARRAY_TYPE`, `MAP_TYPE`, `STRUCT_TYPE` and `USER_DEFINED_TYPE`?

Reply:
> Why do we skip `ARRAY_TYPE`, `MAP_TYPE`, `STRUCT_TYPE` and `USER_DEFINED_TYPE`?

Support this type just convert to string to show.  Should add .

Reply:
I think we should add these types. Hive-3.1.2 also converted these types to strings.
@juliuszsompolski What do you think?

Reply:
How does the client handle it?
If you do
```
val stmt = conn.prepareStatement("SELECT array, map, struct, interval FROM table")
val rs = stmt.executeQuery()
val md = rs.getMetaData()
```
Then what does `md.getColumnType(i)` return for each of these columns?
What type of `rs.getXXX` call should the user use for each of these columns? For the array column, should it be `rs.getArray(i)` or `rs.getString(i)`?
What is the mapping of types returned by md.getColumnType(i), with the getters that should be used for them in rs.getXXX(i)?

Reply:
For getMetadataResult, it truly return ARRAY, MAP, STRUCT.
Return content is organized by HiveResult.toHiveString() method as each's DataType.

Reply:
So if I run this:
```
Class.forName("org.apache.hive.jdbc.HiveDriver")
val jdbc_url = "jdbc:hive2://localhost:10000/default;httpPath=cliservice"
val conn = DriverManager.getConnection(jdbc_url)
val stmt = conn.prepareStatement("select collect_list(id) as arr from range(10)")
val rs = stmt.executeQuery()
val md = rs.getMetaData()
```
and then:
```
scala> md.getColumnType(1)
res2: Int = 2003

scala> java.sql.Types.ARRAY
res3: Int = 2003

scala> rs.next()
res4: Boolean = true

scala> rs.getArray(1)
java.sql.SQLException: Method not supported
  at org.apache.hive.jdbc.HiveBaseResultSet.getArray(HiveBaseResultSet.java:113)
  ... 33 elided

scala> rs.getString(1)
res6: String = [0,1,2,3,4,5,6,7,8,9]
```
Assuming that it's a generic JDBC application, not specific to Hive quirks, how is that application supposed to know that after getting a `getColumnType` of `Types.Array`, it needs to retrieve it with `getString`, and not with `getArray`?

Reply:
@juliuszsompolski 
In hive jdbc method. It will final treat ARRAY, MAP, STRUCT ..etc as String Type:

```
static String columnClassName(Type hiveType, JdbcColumnAttributes columnAttributes)
      throws SQLException {
    int columnType = hiveTypeToSqlType(hiveType);
    switch(columnType) {
      case Types.NULL:
        return "null";
      case Types.BOOLEAN:
        return Boolean.class.getName();
      case Types.CHAR:
      case Types.VARCHAR:
        return String.class.getName();
      case Types.TINYINT:
        return Byte.class.getName();
      case Types.SMALLINT:
        return Short.class.getName();
      case Types.INTEGER:
        return Integer.class.getName();
      case Types.BIGINT:
        return Long.class.getName();
      case Types.DATE:
        return Date.class.getName();
      case Types.FLOAT:
        return Float.class.getName();
      case Types.DOUBLE:
        return Double.class.getName();
      case  Types.TIMESTAMP:
        return Timestamp.class.getName();
      case Types.DECIMAL:
        return BigInteger.class.getName();
      case Types.BINARY:
        return byte[].class.getName();
      case Types.OTHER:
      case Types.JAVA_OBJECT: {
        switch (hiveType) {
          case INTERVAL_YEAR_MONTH_TYPE:
            return HiveIntervalYearMonth.class.getName();
          case INTERVAL_DAY_TIME_TYPE:
            return HiveIntervalDayTime.class.getName();
          default:
            return String.class.getName();
        }
      }
      case Types.ARRAY:
      case Types.STRUCT:
        return String.class.getName();
      default:
        throw new SQLException("Invalid column type: " + columnType);
    }
  }
```

Reply:
OK. We can not fully support these type. Please remove them @AngersZhuuuu 
Thanks @juliuszsompolski for you example.

Reply:
Actually, thanks for explaining it @AngersZhuuuu, and you convinced me that ARRAY, MAP and STRUCT must be included.
```
scala> md.getColumnType(1)
res11: Int = 2003 (== java.sql.Types.ARRAY)
```
but then
```
scala> md.getColumnClassName(1)
res10: String = java.lang.String
```
so that tells to the client that it is actually returned as String, and I should retrieve it as such, either with `rs.getObject(1).asInstance[String]` or as convenient shorthand with `rs.getString(1)`.
It would actually be incorrect to not include Array, Map, Struct, because we do return them in ResultSet schema (through `SparkExecuteStatement.getTableSchema`), so the client can get these type returned, and for any type that can be returned to the client there should be an entry in GetTypeInfo.
We therefore should not include INTERVAL (because we explicitly turn it to String return type after #25277), and not include UNIONTYPE or USER_DEFINED because they don't have any Spark equivalent, but ARRAY, MAP and STRUCT should be there.
Thank you for the explanation :+1:

Reply:
@juliuszsompolski 
```
case Types.OTHER:
      case Types.JAVA_OBJECT: {
        switch (hiveType) {
          case INTERVAL_YEAR_MONTH_TYPE:
            return HiveIntervalYearMonth.class.getName();
          case INTERVAL_DAY_TIME_TYPE:
            return HiveIntervalDayTime.class.getName();
          default:
            return String.class.getName();
        }
      }

``` 
USER_DEFINED  in java.sql.Types is OTHERS, in the convert progress, it's also converted to String , same as ARRAY, MAP, STRUCT. 
Maybe we should add USER_DEFINED.

Reply:
But Spark will never return a USER_DEFINED type.
The current implementation of org.apache.spark.sql.types.UserDefinedType will return the underlying sqlType.simpleString as it's catalogString, so Thriftserver queries will return the underlying type in the schema.
Hence for USER_DEFINED (and UNIONTYPE) the argument is not that they wouldn't potentially work, but that Spark does not use them.

Reply:
> But Spark will never return a USER_DEFINED type.
> The current implementation of org.apache.spark.sql.types.UserDefinedType will return the underlying sqlType.simpleString as it's catalogString, so Thriftserver queries will return the underlying type in the schema.
> Hence for USER_DEFINED (and UNIONTYPE) the argument is not that they wouldn't potentially work, but that Spark does not use them.

Remove it and resolve conflicts.